In [1]:
# Necessary Imports
import pandas as pd
import numpy as np

In [2]:
# Reading Dataframe
data = pd.read_csv('Portland_Crime.csv')

In [3]:
# Data overview
data

Column1                           Address CaseNumber CrimeAgainst  \
0        491800        8800 BLOCK OF SE FLAVEL ST  23-171766     Property   
1        475141           800 BLOCK OF SW 4TH AVE  23-915199     Property   
2        475143           800 BLOCK OF SW 4TH AVE  23-915330     Property   
3        497537       S GAINES ST / S CORBETT AVE  23-171939     Property   
4        490845  7900 BLOCK OF SE ASPEN SUMMIT DR  23-915464     Property   
...         ...                               ...        ...          ...   
499910    19829         4000 BLOCK OF SE BOISE ST  15-143758     Property   
499911    19847     4000 BLOCK OF SE GLADSTONE ST  15-142935     Property   
499912    20090         4100 BLOCK OF SE 82ND AVE  15-143661      Society   
499913    37287                   UNKNOWN ADDRESS  15-654387     Property   
499914    24326        6500 BLOCK OF N LOMBARD ST  15-143369     Property   

              Neighborhood  OccurDate  OccurTime         OffenseCategory  \
0                    Lents  6/30/2023       2100        Larceny Offenses   
1                 Downtown  6/30/2023        730        Larceny Offenses   
2                 Downtown  6/30/2023       1550               Vandalism   
3           South Portland  6/30/2023       1654        Larceny Offenses   
4                    Lents  6/30/2023        930     Motor Vehicle Theft   
...                    ...        ...        ...                     ...   
499910  Creston-Kenilworth   5/1/2015       2121                Burglary   
499911  Creston-Kenilworth   5/1/2015       1900        Larceny Offenses   
499912       Foster-Powell   5/1/2015       1940  Drug/Narcotic Offenses   
499913                 NaN   5/1/2015        825          Fraud Offenses   
499914          Portsmouth   5/1/2015       1100                Burglary   

                                    OffenseType  OpenDataLat  OpenDataLon  \
0                      Theft From Motor Vehicle    45.468464  -122.572201   
1                      Theft From Motor Vehicle    45.517507  -122.677045   
2                                     Vandalism    45.517507  -122.677045   
3                               Purse-Snatching    45.496508  -122.675940   
4                           Motor Vehicle Theft    45.465672  -122.563022   
...                                         ...          ...          ...   
499910                                 Burglary    45.492319  -122.620799   
499911                 Theft From Motor Vehicle    45.493281  -122.620694   
499912                 Drug/Narcotic Violations    45.492416  -122.578840   
499913  False Pretenses/Swindle/Confidence Game          NaN          NaN   
499914                                 Burglary    45.585957  -122.734061   

        OpenDataX  OpenDataY ReportDate  OffenseCount  
0       7670548.0   663706.0  6/30/2023             1  
1       7644142.0   682288.0  6/30/2023             1  
2       7644142.0   682288.0  6/30/2023             1  
3       7644219.0   674627.0  6/30/2023             1  
4       7672876.0   662628.0  6/30/2023             1  
...           ...        ...        ...           ...  
499910  7658313.0   672724.0   5/1/2015             1  
499911  7658349.0   673074.0   5/1/2015             1  
499912  7669070.0   672480.0   5/1/2015             1  
499913        NaN        NaN   5/1/2015             1  
499914  7630223.0   707635.0   5/1/2015             1  

[499915 rows x 15 columns]

The dataframe contains 499915 entries and 15 variables.

In an initial look at the data, the column 'Column1' could be in index column from merging multiple data set, but due to it having incosistent values it will be removed. Yhe column 'OffenseCount' seems to display an int value of 1 for each row and is not a necessary column to keep.

We can spot that there are crimes with an unknown address and location based columns have missing values.
the time in 'OccurTime' is set as military time.

'OccurDate' and 'ReportDate' seem to be identical, this will be investigated further to decide on keeping 'ReportDate'.

'OffenseType' will also be investigating in relation to 'OffenseCategory'.

We will be using Longitude and Latitude for location based research. 'OpenDataY' and 'OpenDataX' can be removed.

In [4]:
# Assigning columns to rename
column_rename_mapping = {
    'OpenDataLat': 'latitude',
    'OpenDataLon': 'longitude'
}

# Applying column renaming to dataframe
data.rename(columns=column_rename_mapping, inplace=True)

In [5]:
# Checking data non null value counts and variable data types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 499915 entries, 0 to 499914
Data columns (total 15 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Column1          499915 non-null  int64  
 1   Address          457678 non-null  object 
 2   CaseNumber       499915 non-null  object 
 3   CrimeAgainst     499915 non-null  object 
 4   Neighborhood     483451 non-null  object 
 5   OccurDate        499915 non-null  object 
 6   OccurTime        499915 non-null  int64  
 7   OffenseCategory  499915 non-null  object 
 8   OffenseType      499915 non-null  object 
 9   latitude         446531 non-null  float64
 10  longitude        446531 non-null  float64
 11  OpenDataX        446531 non-null  float64
 12  OpenDataY        446531 non-null  float64
 13  ReportDate       499915 non-null  object 
 14  OffenseCount     499915 non-null  int64  
dtypes: float64(4), int64(3), object(8)
memory usage: 57.2+ MB


The only data types that will need to be adjusted are the Date related variables

In [6]:
# Converting OccurDate and ReportDate to Datetime
data['OccurDate'] = pd.to_datetime(data['OccurDate'])
data['ReportDate'] = pd.to_datetime(data['ReportDate'])

In [7]:
# Checking for variable missing value count
missing_values = data.isnull().sum()

print(missing_values.sort_values(ascending=False))

latitude           53384
longitude          53384
OpenDataX          53384
OpenDataY          53384
Address            42237
Neighborhood       16464
Column1                0
CaseNumber             0
CrimeAgainst           0
OccurDate              0
OccurTime              0
OffenseCategory        0
OffenseType            0
ReportDate             0
OffenseCount           0
dtype: int64


In [8]:
# Count missing values for longitude and latitude by Neighborhood
missing_counts = data.groupby('Neighborhood')[['longitude', 'latitude']].apply(lambda x: x.isnull().sum()).reset_index()

# Renaming the columns for clarity
missing_counts.columns = ['Neighborhood', 'missing_longitude_count', 'missing_latitude_count']


print(missing_counts)

          Neighborhood  missing_longitude_count  missing_latitude_count
0              Alameda                       72                      72
1          Arbor Lodge                      291                     291
2            Ardenwald                       25                      25
3                Argay                      512                     512
4    Arlington Heights                       18                      18
..                 ...                      ...                     ...
91  West Portland Park                      200                     200
92              Wilkes                      552                     552
93       Woodland Park                       32                      32
94            Woodlawn                      308                     308
95           Woodstock                      279                     279

[96 rows x 3 columns]


We have 96 neighborhood with missing longitude and latitude values.

We can also conclude that the rows will always have both missing according to counts being identical.

The approach to reduce the missing values is locating rows with crime in the same neighborhood with a present longitude and latitude and assigning those values to the rows with the values missing

In [9]:
# Isolating Neighborhood into a list
missing_neighborhoods = missing_counts['Neighborhood'].tolist()

for neighborhood in missing_neighborhoods:
    # Locating a neighborhood in the Dataframe with the same name
    source_row = data[(data['Neighborhood'] == neighborhood) & (~data['longitude'].isnull())].iloc[0]
    
    # Updating the longitude and latitude missing values
    data.loc[data['Neighborhood'] == neighborhood, 'longitude'] = source_row['longitude']
    data.loc[data['Neighborhood'] == neighborhood, 'latitude'] = source_row['latitude']



In [10]:
# Checking for null values
missing_values = data.isnull().sum()

print(missing_values.sort_values(ascending=False))

OpenDataX          53384
OpenDataY          53384
Address            42237
Neighborhood       16464
latitude            9231
longitude           9231
Column1                0
CaseNumber             0
CrimeAgainst           0
OccurDate              0
OccurTime              0
OffenseCategory        0
OffenseType            0
ReportDate             0
OffenseCount           0
dtype: int64


We have greatly reduced the count of missing values for longitude and latitude.

We will keep the data as is because the main focus is on the type of crimes that have been commited and the variable pertaining to it does not have any missing values

In [11]:
# Counter for dates that are different
diff_check = 0

for i in range(len(data)):
    # Checking if the dates are not the same
    if data['OccurDate'][i] != data['ReportDate'][i]:
        diff_check += 1
print('Count of different dates:', diff_check)

Count of different dates: 0


We can conclude that all crimes that have been commited have been reported on the same day, and 'ReportDate' column will not add any value.

In [12]:
# Grouping offense types by categories
offense_categories = data.groupby('OffenseCategory')['OffenseType'].unique()

# Converting results to a dictionary
result = offense_categories.to_dict()

# Displaying results
for category, offense_types in result.items():
    print(f'OffenseCategory: {category}')
    print(f'OffenseTypes: {offense_types}')
    print()

OffenseCategory: Animal Cruelty Offenses
OffenseTypes: ['Animal Cruelty']

OffenseCategory: Arson
OffenseTypes: ['Arson']

OffenseCategory: Assault Offenses
OffenseTypes: ['Aggravated Assault' 'Simple Assault' 'Intimidation']

OffenseCategory: Bribery
OffenseTypes: ['Bribery']

OffenseCategory: Burglary
OffenseTypes: ['Burglary']

OffenseCategory: Counterfeiting/Forgery
OffenseTypes: ['Counterfeiting/Forgery']

OffenseCategory: Drug/Narcotic Offenses
OffenseTypes: ['Drug/Narcotic Violations' 'Drug Equipment Violations']

OffenseCategory: Embezzlement
OffenseTypes: ['Embezzlement']

OffenseCategory: Extortion/Blackmail
OffenseTypes: ['Extortion/Blackmail']

OffenseCategory: Fraud Offenses
OffenseTypes: ['Identity Theft' 'Wire Fraud' 'False Pretenses/Swindle/Confidence Game'
 'Credit Card/ATM Fraud' 'Impersonation' 'Hacking/Computer Invasion'
 'Welfare Fraud']

OffenseCategory: Gambling Offenses
OffenseTypes: ['Operating/Promoting/Assisting Gambling']

OffenseCategory: Homicide Offenses


After careful observation it seems that there isn't great value in keeping OffenseType unless the analysis is to investigate specific crimes from each category where this analysis is not. The 'OffenseTypes' column will be removed as a result

In [19]:
# Removing Unnecessary columns
cols_to_remove = ['Column1', 'ReportDate', 'OffenseCount', 'OpenDataX', 'OpenDataY', 'OffenseType']
data = data.drop(columns = cols_to_remove)
data

Address CaseNumber CrimeAgainst  \
0             8800 BLOCK OF SE FLAVEL ST  23-171766     Property   
1                800 BLOCK OF SW 4TH AVE  23-915199     Property   
2                800 BLOCK OF SW 4TH AVE  23-915330     Property   
3            S GAINES ST / S CORBETT AVE  23-171939     Property   
4       7900 BLOCK OF SE ASPEN SUMMIT DR  23-915464     Property   
...                                  ...        ...          ...   
499910         4000 BLOCK OF SE BOISE ST  15-143758     Property   
499911     4000 BLOCK OF SE GLADSTONE ST  15-142935     Property   
499912         4100 BLOCK OF SE 82ND AVE  15-143661      Society   
499913                   UNKNOWN ADDRESS  15-654387     Property   
499914        6500 BLOCK OF N LOMBARD ST  15-143369     Property   

              Neighborhood  OccurDate  OccurTime         OffenseCategory  \
0                    Lents 2023-06-30       2100        Larceny Offenses   
1                 Downtown 2023-06-30        730        Larceny Offenses   
2                 Downtown 2023-06-30       1550               Vandalism   
3           South Portland 2023-06-30       1654        Larceny Offenses   
4                    Lents 2023-06-30        930     Motor Vehicle Theft   
...                    ...        ...        ...                     ...   
499910  Creston-Kenilworth 2015-05-01       2121                Burglary   
499911  Creston-Kenilworth 2015-05-01       1900        Larceny Offenses   
499912       Foster-Powell 2015-05-01       1940  Drug/Narcotic Offenses   
499913                 NaN 2015-05-01        825          Fraud Offenses   
499914          Portsmouth 2015-05-01       1100                Burglary   

         latitude   longitude  
0       45.468464 -122.572201  
1       45.517507 -122.677045  
2       45.517507 -122.677045  
3       45.496508 -122.675940  
4       45.468464 -122.572201  
...           ...         ...  
499910  45.494021 -122.625645  
499911  45.494021 -122.625645  
499912  45.483952 -122.578995  
499913        NaN         NaN  
499914  45.582129 -122.721192  

[499915 rows x 9 columns]

In [20]:
data.to_csv('updated_Portland_Crime.csv', index=False)